Update EIP-3540: Move to Review (#8152)
ethereum/EIPs: EIPS/eip-3540.md

type: Standards Track
category: Core
created: 2021-03-16

> requires: 3541, 3860, 4750, 5450, 7620

Reviewer: I'd remove the circular dependencies: 4750, 5450, 7620.

The opcode `0xEF` is currently an undefined instruction, therefore: *It pops no stack items and pushes no stack items, and it causes an exceptional abort when executed.* This means legacy *initcode* or already deployed legacy *code* starting with this instruction will continue to abort execution.

Unless otherwised specified, all integers are encoded in big-endian byte order.

### Code validation

We introduce *code validation* for new contract creation. To achieve this, we define a format called EVM Object Format (EOF), containing a version indicator, and a ruleset of validity tied to a given version.

Legacy code is not affected by EOF code validation.

Code validation is performed during contract creation, and is elaborated on in separate EIPs.
The EOF format itself and its formal validation are described in the following sections.

### Container specification

EOF container is a binary format with the capability of providing the EOF version number and a list of EOF sections.

The container starts with the EOF prefix:

| description | length   | value      |                    |
|-------------|----------|------------|--------------------|

> | magic       | 2-bytes  | 0xEF00     |                    |

Reviewer 2: Can we add a ref to The MAGIC section below?

Author: As mentioned elsewhere, I think this is a tiny bit non-standard to link to the Rationale section bits. If done scrupulously for all the pieces of the spec, would overwhelm the spec itself.

### Remarks

If code starts with the `MAGIC`, it is considered to be EOF formatted, otherwise it is considered to be *legacy* code. For clarity, the `MAGIC` together with a version number *n* is denoted as the *EOFn prefix*, e.g. *EOF1 prefix*.

EOF-formatted contracts are created using new instructions `CREATE3`, `CREATE4` and `RETURNCONTRACT`. They are introduced in a separate EIP-???? (TBD, refer to the [Megaspec doc](https://github.com/ipsilon/eof/blob/main/spec/eof.md) until then).

The opcode `0xEF` is currently an undefined instruction, therefore: *It pops no stack items and pushes no stack items, and it causes an exceptional abort when executed.* This means legacy *initcode* or already deployed legacy *code* starting with this instruction will continue to abort execution.

Unless otherwised specified, all integers are encoded in big-endian byte order.

### Code validation

> We introduce *code validation* for new contract creation, which EOF formatted code will undergo. To achieve this, we define a format called EVM Object Format (EOF), containing a version indicator, and a ruleset of validity tied to a given version.

Reviewer 3: I think it's fine to leave like before, in you version "EOF" is used before being defined in next sentence.
```suggestion
We introduce *code validation* for new contract creation. To achieve this, we define a format called EVM Object Format (EOF), containing a version indicator, and a ruleset of validity tied to a given version.
```

Legacy code is not affected by EOF code validation.

> Code validation is performed during `CREATE4` instruction, and is elaborated on in [EIP-3670](./eip-3670.md) and an upcomming contract creation EIP-???? (TBD, refer to the [Megaspec doc](https://github.com/ipsilon/eof/blob/main/spec/eof.md) until then)

Reviewer 3: ```suggestion
Code validation is performed during `CREATE4` instruction, and is elaborated on in [EIP-3670](./eip-3670.md) and an upcoming contract creation EIP-???? (TBD, refer to the [Megaspec doc](https://github.com/ipsilon/eof/blob/main/spec/eof.md) until then)
```

The EOF format itself and its formal validation are described in the following sections.

### Container specification

EOF container is a binary format with the capability of providing the EOF version number and a list of EOF sections.

The container starts with the EOF prefix:

| description | length   | value      |                    |
|-------------|----------|------------|--------------------|
| magic       | 2-bytes  | 0xEF00     |                    |
| version     | 1-byte   | 0x01–0xFF  | EOF version number |

The EOF prefix is followed by at least one section header. Each section header contains two fields, `section_kind` and bytes describing the section of structure defined specifically for each kind.

| description         | length  | value         |                   |
|---------------------|---------|---------------|-------------------|
| section_kind        | 1-byte  | 0x01–0xFF     | `uint8`           |

> | section description | dynamic | n/a           | `uint8+`          |

Reviewer 3: Not sure why this part was changed, I think previous general description still applies?

Author: It doesn't - for code and containers sections it is not a 2-byte size anymore.

Reviewer 3: `section_size` applied only to single-size sections (type and data), and `section_size_list` applied to code sections and now should also apply to container sections. The paragraph above described this well:

> The EOF header is followed by at least one section header. Each section header contains two fields, `section_kind` and either `section_size` or `section_size_list`, depending on the kind. `section_size_list` is a list of size values when multiple sections of this kind are allowed.

Author: Oh, so `_list` here included the number of items as the first `uint16` :facepalm: . I somehow misread this and the "or", and it seemed completely out-of-date

Nevertheless. I thought we wanted to make the general EOF header format more liberally defined, per the rationale included below:

> - Whether or not to have more structure to the container header for all EOF versions to follow. In order to allow future formats optimized for chunking and merkleization (verkleization) it was decided to keep it generic and specify the structure only for a specific EOF version.

This would invite putting offsets in, rather than sizes.

Should we revert this and require a specific structure for all versions of EOF?

Author: Agreed on the call to revert, https://github.com/ethereum/EIPs/pull/8152/commits/7e4c93bfc507dd84a59cb6dfe02f8a6711e84258 (I also took the liberty to be more specific about how the list is encoded)

| data_section      | variable | n/a           | arbitrary sequence of bytes                                                                |

See [EIP-4750](./eip-4750.md) for more information on the type section content.

**NOTE**: A special value of `outputs` being `0x80` is designated to denote non-returning functions as defined in a separate EIP.

#### EOF version 1 validation rules

The following validity constraints are placed on the container format:

- `version` must be `0x01`
- `types_size` is divisible by `4`
- the number of code sections must be equal to `types_size / 4`
- the number of code sections must be greater than `0` and not exceed `1024`
- `code_size` may not be `0`
- the number of container sections must not exceed `256`. The number of container sections may not be `0`, if declared in the header
- `container_size` may not be `0`
- data section is mandatory, but `data_size` may be `0`
- data body length may be shorter than `data_size` for a not yet deployed container

### Changes to execution semantics

For an EOF contract:

- Execution starts at the first byte of code section 0, and PC is set to 0.

> - `PC` is scoped to the executing code section

Reviewer 3: Probably this also can be removed.

Author: right, but isn't it a remark on how the context variable PC works, rather than opcode? I never was really sure. If you think opcode - let's remove. Actually, I think EIP-4750 elaborates enough on how PC works as a context variable, so I guess we can remove anyway.

Reviewer 3: I thought so too at first, but it's next to the list of what happens to other opcodes, so very much looks like about opcode too.

Author: 088f5e25 a final check on this pls. I think this leaves PC out completely as irrelevant, making 3540 define how EOF would have worked without CALLF/RETF (only 0th code section executes), which I think is coherent.

1. `version` MUST NOT be `0`.[^1](#eof-version-range-start-with-1)
2. `section_kind` MUST NOT be `0`. The value `0` is reserved for *section headers terminator byte*.
3. There MUST be at least one section (and therefore section header).
5. Stray bytes outside of sections MUST NOT be present. This includes trailing bytes after the last section.

### EOF version 1

EOF version 1 is made up of several EIPs, including this one, as enumerated in the EOF Meta EIP-???? (TBD). Some values in this specification are only discussed briefly. To understand the full scope of EOF, it is necessary to review each EIP in-depth.

#### Container

The EOF version 1 container consists of a `header` and `body`.

```
container := header, body
header := 
    magic, version, 

>     kind_types, types_size, 

Reviewer 3: ```suggestion
    kind_type, type_size, 
```

    kind_code, num_code_sections, code_size+,
    [kind_container, num_container_sections, container_size+,]
    kind_data, data_size,
    terminator
body := types_section, code_section+, container_section*, data_section
types_section := (inputs, outputs, max_stack_height)+
```

> _note: `,` is a concatenation operator, `+` should be interpreted as "one or more" of the preceding item, and `*` should be interpreted as "zero or more" of the preceding item._

Reviewer 3: Linter requires to change underscore to asterisk here.

> _note: `,` is a concatenation operator, `+` should be interpreted as "one or more" of the preceding item, and `*` should be interpreted as "zero or more" of the preceding item._

Reviewer 3: Add also that square brackets mean optionality (0 or 1).

#### Header

| name                   | length   | value         | description                                                                             |
|------------------------|----------|---------------|-----------------------------------------------------------------------------------------|
| magic                  | 2 bytes  | 0xEF00        |                                                                                         |
| version                | 1 byte   | 0x01          | EOF version                                                                             |

> | kind_types             | 1 byte   | 0x01          | kind marker for types size section                                                      |

Author: got it from Megaspec. In general, once dust settles here, I'll port the refreshed format of the tables back to there :)

> | types_size             | 2 bytes  | 0x0004-0xFFFF | 16-bit unsigned big-endian integer denoting the length of the type section content      |

Author: Makes sense, but this is taken from the megaspec, so the values there incorrect? Are we 100% sure the change wasn't intended?

| kind_code              | 1 byte   | 0x02          | kind marker for code size section                                                       |

> | num_code_sections      | 2 bytes  | 0x0001-0xFFFF | 16-bit unsigned big-endian integer denoting the number of the code sections             |

Reviewer 3: ```suggestion
| num_code_sections      | 2 bytes  | 0x0001-0x0400 | 16-bit unsigned big-endian integer denoting the number of the code sections             |
```

| code_size              | 2 bytes  | 0x0001-0xFFFF | 16-bit unsigned big-endian integer denoting the length of the code section content      |
| kind_container         | 1 byte   | 0x03          | kind marker for container size section                                                  |
| num_container_sections | 2 bytes  | 0x0001-0x00FF | 16-bit unsigned big-endian integer denoting the number of the container sections        |
| container_size         | 2 bytes  | 0x0001-0xFFFF | 16-bit unsigned big-endian integer denoting the length of the container section content |
| kind_data              | 1 byte   | 0x04          | kind marker for data size section                                                       |
| data_size              | 2 bytes  | 0x0000-0xFFFF | 16-bit unsigned big-endian integer denoting the length of the data section content (*)  |
| terminator             | 1 byte   | 0x00          | marks the end of the header                                                             |

> (*) For not yet deployed containers this can be more than the actual content, see Data Section Lifecycle section in the contract creation EIP-???? (TBD + link).

Reviewer 3: ```suggestion
(*) For not yet deployed containers this can be greater than the actual content length, see Data Section Lifecycle section in the contract creation EIP-???? (TBD + link).
```

#### Body

| name              | length   | value         | description                                                                           |
|-------------------|----------|---------------|---------------------------------------------------------------------------------------|
| types_section     | variable | n/a           | stores code section metadata                                                          |
| inputs            | 1 byte   | 0x00-0x7F     | number of stack elements the code section consumes                                    |
| outputs           | 1 byte   | 0x00-0x80     | number of stack elements the code section returns or 0x80 for non-returning functions |

> | max_stack_height  | 2 bytes  | 0x0000-0x03FF | maximum number of elements ever placed onto the stack by the code section             |

Reviewer 3: ```suggestion
| max_stack_height  | 2 bytes  | 0x0000-0x03FF | maximum number of elements ever placed onto the operand stack by the code section             |
```

> | code_section      | variable | n/a           | arbitrary sequence of bytes                                                           |
> | container_section | variable | n/a           | arbitrary sequence of bytes                                                           |

Reviewer 3: ```suggestion
| code_section      | variable | n/a           | arbitrary bytecode                                                         |
| container_section | variable | n/a           | arbitrary EOF-formatted container                                                           |
```

| types_section     | variable | n/a           | stores code section metadata                                                          |
| inputs            | 1 byte   | 0x00-0x7F     | number of stack elements the code section consumes                                    |
| outputs           | 1 byte   | 0x00-0x80     | number of stack elements the code section returns or 0x80 for non-returning functions |
| max_stack_height  | 2 bytes  | 0x0000-0x03FF | maximum number of elements ever placed onto the stack by the code section             |
| code_section      | variable | n/a           | arbitrary sequence of bytes                                                           |
| container_section | variable | n/a           | arbitrary sequence of bytes                                                           |
| data_section      | variable | n/a           | arbitrary sequence of bytes                                                           |

See [EIP-4750](./eip-4750.md) for more information on the type section content.
See [EIP-6206](./eip-6206.md) for more information on non-returning functions.

#### EOF version 1 validation rules

> 1. In addition to general validation rules above, EOF version 1 bytecode conforms to the rules specified below:

Reviewer 3: This part was mandating exact order of the sections, we are kind of missing it now, but maybe the formula in `Container` section is enough.

Author: yea, I assumed it's enough and much cleaner than this

> - `container_size` may not be 0, but container sections are optional

Reviewer 3: I would also add
```
- data section is mandatory, but `data_size` may be 0
```

Reviewer 3: Also maybe something like
```
- data body length may be shorter than `data_size`
```

> - the number of code sections must not exceed 1024
> - `code_size` may not be 0
> - the number of container sections must not exceed 256
> - `container_size` may not be 0, but container sections are optional
> - data section is mandatory, but `data_size` may be 0
> - data body length may be shorter than `data_size` for a not yet deployed container

Reviewer 3: ```suggestion
- the number of code sections must not exceed `1024`
- `code_size` may not be `0`
- the number of container sections must not exceed `256`
- `container_size` may not be `0`, but container sections are optional
- data section is mandatory, but `data_size` may be `0`
- data body length may be shorter than `data_size` for a not yet deployed container
```

> - the total size of not yet deployed container might be up to `data_size` lower than the above values due to how the data section is rewritten and resized during creation

Reviewer 3: I think these are not rules and should be removed

> - Execution starts at the first byte of code section 0, and `pc` is set to 0.

Reviewer 3: Linter requirement
```suggestion

- Execution starts at the first byte of code section 0, and `pc` is set to 0.
```

> - `CODESIZE`, `CODECOPY`, `EXTCODESIZE`, `EXTCODECOPY`, `EXTCODEHASH`, `GAS` are rejected by validation in EOF contracts, with no replacements

Reviewer 3: I think we should either remove this, or have a full list of deprecated instructions. But it should be in EIP-3670, so maybe remove.

Author: After some deliberation, I decided to have the deprecations accompany the replacement definitions wherever possible, and here list all the deprecations without replacements. The EIPs also kinda followed this model, so retaining this should give a smaller diff. But mostly, I think it reads easier, if the legacy and EOF logical counterparts are nearby.

But that's not a strong opinion, I spent quite a lot of time juggling these alternatives, all have pros and cons

| inputs            | 1 byte   | 0x00-0x7F     | number of stack elements the code section consumes                                         |
| outputs           | 1 byte   | 0x00-0x7F     | number of stack elements the code section returns                                          |
| max_stack_height  | 2 bytes  | 0x0000-0x03FF | maximum number of elements ever placed onto the operand stack by the code section          |
| code_section      | variable | n/a           | arbitrary bytecode                                                                         |
| container_section | variable | n/a           | arbitrary EOF-formatted container                                                          |
| data_section      | variable | n/a           | arbitrary sequence of bytes                                                                |

See [EIP-4750](./eip-4750.md) for more information on the type section content.

#### EOF version 1 validation rules

The following validity constraints are placed on the container format:

- `version` must be `0x01`
- `types_size` is divisible by `4`
- the number of code sections must be equal to `types_size / 4`

> - the number of code sections must not exceed `1024`

Reviewer 3: It's mentioned in the table, but maybe worth to mention here explicitly that `num_code_sections > 0`.

- `code_size` may not be `0`

> - the number of container sections must not exceed `256`

Reviewer 3: Maybe worth to mention here explicitly that `num_container_sections > 0`, if they are present (you can't have container section declared with 0 number of containers)

| kind_data              | 1 byte   | 0x04          | kind marker for data size section                                                                            |
| data_size              | 2 bytes  | 0x0000-0xFFFF | 16-bit unsigned big-endian integer denoting the length of the data section content (*)                       |
| terminator             | 1 byte   | 0x00          | marks the end of the header                                                                                  |

(*) For not yet deployed containers this can be greater than the actual content length.

#### Body

| name              | length   | value         | description                                                                                |
|-------------------|----------|---------------|--------------------------------------------------------------------------------------------|
| types_section     | variable | n/a           | stores code section metadata                                                               |
| inputs            | 1 byte   | 0x00-0x7F     | number of stack elements the code section consumes                                         |

> | outputs           | 1 byte   | 0x00-0x80     | number of stack elements the code section returns or 0x80 for non-returning functions      |

Reviewer 3: I'm not a fan of this, because without it it seems you can implement the format according to EIP and then pass 0 tests.

Maybe I'm wrong, because you wouldn't pass anything without implementing the rest of validation anyway.

> | outputs           | 1 byte   | 0x00-0x80     | number of stack elements the code section returns or 0x80 for non-returning functions      |

Author: Following the convention we seem to have adopted recently with the other EIPs, I'll leave a note that one of `outputs`' special values has already been assigned in a separate EIPs, or sth like that.

| container_section | variable | n/a           | arbitrary EOF-formatted container                                                          |
| data_section      | variable | n/a           | arbitrary sequence of bytes                                                                |

See [EIP-4750](./eip-4750.md) for more information on the type section content.

**NOTE**: A special value of `outputs` being `0x80` is designated to denote non-returning functions as defined in a separate EIP.

#### EOF version 1 validation rules

The following validity constraints are placed on the container format:

- `version` must be `0x01`
- `types_size` is divisible by `4`
- the number of code sections must be equal to `types_size / 4`
- the number of code sections must be greater than `0` and not exceed `1024`
- `code_size` may not be `0`
- the number of container sections must not exceed `256`. The number of container sections may not be `0`, if declared in the header
- `container_size` may not be `0`
- data section is mandatory, but `data_size` may be `0`
- data body length may be shorter than `data_size` for a not yet deployed container

> 

Reviewer 2: This list is different than [Container Validation](https://github.com/ipsilon/eof/blob/main/spec/eof.md#container-validation)

### Changes to execution semantics

For an EOF contract:

- Execution starts at the first byte of code section 0, and PC is set to 0.
- `PC` is scoped to the executing code section
- `CODESIZE`, `CODECOPY`, `EXTCODESIZE`, `EXTCODECOPY`, `EXTCODEHASH`, `GAS` are rejected by validation in EOF contracts, with no replacements

> - `CALL`, `DELEGATECALL`, `STATICCALL` are rejected by validation in EOF contracts, replacement instructions to be introduced in a separate EIP.

Reviewer 2: Can we add a reference?

- (EOF1) Whether to encode section sizes as fixed 16-bit values or some kind of variable length field (e.g. LEB128). We have opted for fixed size, because it simplifies client implementations, and 16-bit seems enough, because of the currently exposed code size limit of 24576 bytes (see [EIP-170](./eip-170.md) and [EIP-3860](./eip-3860.md)). Should this be limiting in the future, a new EOF version could change the format. Besides simplifying client implementations, not using LEB128 also greatly simplifies on-chain parsing.
- Whether or not to have more structure to the container header for all EOF versions to follow. In order to allow future formats optimized for chunking and merkleization (verkleization) it was decided to keep it generic and specify the structure only for a specific EOF version.

### Data-only contracts

(TBD: Moved to an upcoming section in the [EIP-7480](./eip-7480.md))

> ### `pc` starts with 0 at the code section

Reviewer 3: Why did you make PC lowercase?

- Whether to have a header terminator or to encode `number_of_sections` or `total_size_of_headers`. Both raise the question of how large of a value these fields should be able to hold. A terminator byte seems to avoid the problem of choosing a size which is too small without any perceptible downside, so it is the path taken.
- (EOF1) Whether to encode section sizes as fixed 16-bit values or some kind of variable length field (e.g. LEB128). We have opted for fixed size, because it simplifies client implementations, and 16-bit seems enough, because of the currently exposed code size limit of 24576 bytes (see [EIP-170](./eip-170.md) and [EIP-3860](./eip-3860.md)). Should this be limiting in the future, a new EOF version could change the format. Besides simplifying client implementations, not using LEB128 also greatly simplifies on-chain parsing.
- Whether or not to have more structure to the container header for all EOF versions to follow. In order to allow future formats optimized for chunking and merkleization (verkleization) it was decided to keep it generic and specify the structure only for a specific EOF version.

### Data-only contracts

> Moved to the section [Lack of `EXTDATACOPY` in EIP-7480](./eip-7480#lack-of-extdatacopy.md).

Reviewer 3: I guess
```suggestion
Moved to the section [Lack of `EXTDATACOPY` in EIP-7480](./eip-7480.md#lack-of-extdatacopy).
```

- Whether to have a header terminator or to encode `number_of_sections` or `total_size_of_headers`. Both raise the question of how large of a value these fields should be able to hold. A terminator byte seems to avoid the problem of choosing a size which is too small without any perceptible downside, so it is the path taken.
- (EOF1) Whether to encode section sizes as fixed 16-bit values or some kind of variable length field (e.g. LEB128). We have opted for fixed size, because it simplifies client implementations, and 16-bit seems enough, because of the currently exposed code size limit of 24576 bytes (see [EIP-170](./eip-170.md) and [EIP-3860](./eip-3860.md)). Should this be limiting in the future, a new EOF version could change the format. Besides simplifying client implementations, not using LEB128 also greatly simplifies on-chain parsing.
- Whether or not to have more structure to the container header for all EOF versions to follow. In order to allow future formats optimized for chunking and merkleization (verkleization) it was decided to keep it generic and specify the structure only for a specific EOF version.

### Data-only contracts

> Moved to the section [Lack of `EXTDATACOPY` in EIP-7480](./eip-7480.md#lack-of-extdatacopy).

Reviewer 2: I think we don't have to mentioned that it's moved. We can just add a reference to "Lack of ..."

> ### `PC` starts with 0 at the code section

Reviewer 3: I think this entire paragraph is outdated, it explained why PC opcode returns what it does, when it was still allowed. It is irrelevant now.

## Test Cases

> ### Container validation rules

Reviewer 3: I would just remove Test Cases section and focus on providing state tests instead.

(TBD)

### Contract execution

> - EOF0 `EXTCODECOPY` to an EOF1 contract copies 0 bytes

Author: yup, leftovers again :(. But I'm removing test cases anyway

- EOF0 `EXTCODEHASH` to an EOF1 contract returns hash of `0xEF00`
- EOF0 `EXTCODESIZE` to an EOF1 contract returns 2

> - EOF0 `DELEGATECALL2` to EOF1 code succeeds

Reviewer 3: I think we decided that new calls are only for EOF?
```suggestion
- EOF0 `DELEGATECALL` to EOF1 code succeeds
```

Author: correct, I updated too much here
